library/std/src/rt.rs

```diff
@@ -25,17 +25,35 @@ fn lang_start_internal(
     argc: isize,
     argv: *const *const u8,
 ) -> isize {
-    use crate::panic;
-    use crate::sys_common;
+    use crate::{mem, panic, sys, sys_common};
+    // Guard against the code called by this function from unwinding outside of the Rust-controlled
+    // code, which is UB. This is a requirement imposed by a combination of how the
+    // `#[lang="start"]` attribute is implemented as well as by the implementation of the panicking
+    // mechanism itself.
+    //
+    // There are a couple of instances where unwinding can begin. First is inside of the
+    // `rt::init`, `rt::cleanup` and similar functions controlled by libstd. In those instances a
+    // panic is a libstd implementation bug. A quite likely one too, as there isn't any way to
+    // prevent libstd from accidentally introducing a panic to these functions. Another is from
+    // user code, as described in e.g. issue #86030.
+    let result = panic::catch_unwind(move || {
+        // SAFETY: Only called once during runtime initialization.
+        unsafe { sys_common::rt::init(argc, argv) };
 
-    // SAFETY: Only called once during runtime initialization.
-    unsafe { sys_common::rt::init(argc, argv) };
+        let exit_code = panic::catch_unwind(main);
 
-    let exit_code = panic::catch_unwind(main);
-
-    sys_common::rt::cleanup();
-
-    exit_code.unwrap_or(101) as isize
+        sys_common::rt::cleanup();
+        exit_code.unwrap_or(101) as isize
```

Comment:
:thought_balloon: at this location we invoke user code after the runtime cleanup has been executed. Is moving the drop here to above the cleanup something we want to do?

Reply:
I would guess so.

Reply:
Yes, that'd be good.